internal/staging/primitives/pow.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package primitives
+
+import (
+	"fmt"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/internal/staging/primitives/uint256"
+)
+
+// DiffBitsToUint256 converts the compact representation used to encode
+// difficulty targets to an unsigned 256-bit integer.  The representation is
+// similar to IEEE754 floating point numbers.
+//
+// Like IEEE754 floating point, there are three basic components: the sign,
+// the exponent, and the mantissa.  They are broken out as follows:
+//
+//	* the most significant 8 bits represent the unsigned base 256 exponent
+//	* bit 23 (the 24th bit) represents the sign bit
+//	* the least significant 23 bits represent the mantissa
+//
+//	-------------------------------------------------
+//	|   Exponent     |    Sign    |    Mantissa     |
+//	-------------------------------------------------
+//	| 8 bits [31-24] | 1 bit [23] | 23 bits [22-00] |
+//	-------------------------------------------------
+//
+// The formula to calculate N is:
+//	N = (-1^sign) * mantissa * 256^(exponent-3)
+//
+// Note that this encoding is capable of representing negative numbers as well
+// as numbers much larger than the maximum value of an unsigned 256-bit integer.
+// However, it is only used in Decred to encode unsigned 256-bit integers which
+// represent difficulty targets, so rather than using a much less efficient
+// arbitrary precision big integer, this implementation uses an unsigned 256-bit
+// integer and returns flags to indicate whether or not the encoding was for a
+// negative value and/or overflows a uint256 to enable proper error detection
+// and stay consistent with legacy code.
+func DiffBitsToUint256(bits uint32) (n uint256.Uint256, isNegative bool, overflows bool) {
+	// Extract the mantissa, sign bit, and exponent.
+	mantissa := bits & 0x007fffff
+	isSignBitSet := bits&0x00800000 != 0
+	exponent := bits >> 24
+
+	// Nothing to do when the mantissa is zero as any multiple of it will
+	// necessarily also be 0 and therefore it can never be negative of overflow.
```

Comment:
Fixed.

I also reworded the comments in the following sections a bit to clarify things better since they were just copied from old code and I noticed they probably weren't super clear without already knowing that history because it talks about bytes which are no longer used.